### PR TITLE
Adding transaction types and aborted transactions/solutions to metrics

### PR DESCRIPTION
**Pull request by miazn**
## Motivation
It's useful to have more granular metrics as to what types of transactions are included in blocks, as well as aborted transactions and solutions.

This pr adds 6 new metrics:
```
1. Accepted Deploy
2. Accepted Execute
3. Rejected Deploy
4. Rejected Execute
5. Aborted Transactions
6. Aborted Solutions
```

I did not yet remove the existing confirmed transactions (which we can get by summing the accepted deploys and executes), but that would probably be best to avoid duplicate data (can sum in post processing for analysis).
## Test Plan
Screenshot 2024-04-11 at 3 42 45 PM
Still testing to get all types recorded

### CODE REVIEW

**Comment by miazn** [comment]
@ljedrz is there a more optimal way to do the enum matching?

**Comment by ljedrz** [comment]
Was this found to be slow? Enum matching on its own shouldn't be a big deal, it's the mutable borrows that e.g. disallow parallel processing; if the iteration needs to be faster, the counters can be made atomic, then the `_mut`s can be dropped from the `get`s, and the iteration can be parallelized with `rayon`.

File: node/bft/ledger-service/src/traits.rs
Comment (iamalwaysuncomfortable) [inline node/bft/ledger-service/src/traits.rs]: Just as a general nit, traits are interfaces and when other Rust code needs to accept a trait parameter elsewhere, it expects the concrete type that implements it to have it implemented as well. 

Putting a feature flag in a trait basically makes other implementors of that trait need to also implement a feature flag in their implementation further down the line which can be lead to a lot of problems. Thus in general it's good NOT to put feature flags in traits for this reason unless it's somehow absolutely necessary. (There is another feature flag in the trait above, but this is now adding a separate one) 

In the concrete implementation of this above, the &self parameter isn't used and only block data from the parameter `block` input itself is ever used (i.e. you don't need to access &self at all). Therefore you don't need it as part of the trait and can make the `update_block_metrics` call from either a standalone function or a method on another object.

Luckily though this should be a simple change!

File: node/bft/ledger-service/src/ledger.rs
Comment (iamalwaysuncomfortable) [inline node/bft/ledger-service/src/ledger.rs]: To track Aborted Transactions & Aborted Solutions. Block has two methods:

`aborted_solution_ids` and `aborted_transaction_ids` - you can use these two methods to count the aborted solutions & transactoins.

**Review by iamalwaysuncomfortable — DISMISSED** [review]
Left a comment about how to track aborted solutions & transactions.

The one major nit is to remove the metrics method from the ledger trait. Reasons why are explained int he comments.

Thanks for working to contribute this to the node software. This will make the life of all node operators easier once it's merged!

File: node/bft/ledger-service/src/traits.rs
Comment (miazn) [reply node/bft/ledger-service/src/traits.rs]: Yeah I was initially trying to make it a standalone function but for some reason it would not compile. Trying this again now

File: node/bft/ledger-service/src/ledger.rs
Comment (miazn) [reply node/bft/ledger-service/src/ledger.rs]: I used those in the function already, they are the last two lines

File: node/bft/ledger-service/src/traits.rs
Comment (miazn) [reply node/bft/ledger-service/src/traits.rs]: Changed :D

File: node/bft/ledger-service/src/ledger.rs
Comment (iamalwaysuncomfortable) [reply node/bft/ledger-service/src/ledger.rs]: Nice! Let me take a second look.

File: node/bft/ledger-service/src/lib.rs
Comment (iamalwaysuncomfortable) [inline node/bft/ledger-service/src/lib.rs]: My personal opinion on where this should go is in the `metrics` crate. This way if there's any question about the metrics collection logic, we have a singular place to look. My fear is that if we end up having a lot of metrics collection logic beyond this, it may get hard to keep track of it - and keeping as much as is possible in the metrics crate will help us keep track of our metrics logic.

Does anyone have any other opinions on where this function should live?

File: node/metrics/snarkOS-grafana.json
Comment (miazn) [inline node/metrics/snarkOS-grafana.json:206]: Changed these names just to reflect changes in https://github.com/AleoHQ/snarkOS/pull/3210

File: node/bft/ledger-service/src/lib.rs
Comment (howardwu) [inline node/bft/ledger-service/src/lib.rs]: `cfg_iter!`?

Comment (iamalwaysuncomfortable) [reply node/bft/ledger-service/src/lib.rs]: @miazn for context `cfg_iter!` is a Rust macro within SnarkVM that will iterate through a collection either in parallel or serial depending on the feature flag. 

A usage example is can be found here [here](https://github.com/AleoHQ/snarkVM/blob/d48f6fb7d7ca36d134407baebcc59952fe58c546/ledger/block/src/verify.rs#L245-L255). 

The actual source code behind this macro is [here](https://github.com/AleoHQ/snarkVM/blob/d48f6fb7d7ca36d134407baebcc59952fe58c546/utilities/src/parallel.rs#L88-L98).

Since there exists a `serial` flag that will disable parallel iteration. It's important to use this macro to ensure that this code will still run if the `serial` flag is enabled.

Comment (miazn) [reply node/bft/ledger-service/src/lib.rs]: ah gotcha, wasn't privy to that -- changed :D

File: node/bft/ledger-service/src/lib.rs
Comment (iamalwaysuncomfortable) [reply node/bft/ledger-service/src/lib.rs]: Pending other reviewer's comments on where this function should live, this PR has my approval.

File: node/bft/ledger-service/Cargo.toml
Comment (howardwu) [inline node/bft/ledger-service/Cargo.toml]: `optional = true` and add `serial`?

Comment (miazn) [reply node/bft/ledger-service/Cargo.toml]: done!

**Comment by miazn** [comment]
after discussing with @iamalwaysuncomfortable , we decided it might make the most sense to move the metrics related functions to the actual metrics crate. This makes it possible to import `rayon` optionally when `metrics` is enabled and `serial` is disabled , as well as the added bonus of cleaning up protocol logic code

**Comment by howardwu** [comment]
thanks for making the change!

File: node/bft/ledger-service/Cargo.toml
Comment (howardwu) [inline node/bft/ledger-service/Cargo.toml]: This should be removed.

File: node/bft/ledger-service/Cargo.toml
Comment (howardwu) [inline node/bft/ledger-service/Cargo.toml]: This should be removed.

File: node/metrics/Cargo.toml
Comment (howardwu) [inline node/metrics/Cargo.toml]: nit: this should be reordered for alphabetical ordering

File: node/metrics/Cargo.toml
Comment (raychu86) [inline node/metrics/Cargo.toml]:
```suggestion
version = "0.3"

```

nit: new line at end of file

File: node/metrics/src/lib.rs
Comment (howardwu) [inline node/metrics/src/lib.rs]: nit: please fix formatting and import ordering

File: node/metrics/src/lib.rs
Comment (howardwu) [inline node/metrics/src/lib.rs]:
```suggestion
pub use names::*;

// Re-export the snarkVM metrics.
pub use snarkvm::metrics::*;

#[cfg(not(feature = "serial"))]
use rayon::prelude::*;

use parking_lot::Mutex;
```

**Review by howardwu — APPROVED** [review]

**Pull request merged**
